samples: bluetooth: add CTS client sample

leewkb4567 opened the pull request:
Add Peripheral CTS client sample to demonstrate how to use the Current Time Service Client.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(cts_c, CONFIG_BT_GATT_CTS_C_LOG_LEVEL);
+
+/** The lowest valid Current Time year is the year when the Western calendar was introduced. */
```
`/**` ->  `/*` This file isn't parsed by doxygen

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+
+/** The lowest valid Current Time year is the year when the Western calendar was introduced. */
+#define CTS_YEAR_MIN 1582
+/** The highest possible Current Time year. */
```
`/**` -> `*/`

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+#define CTS_YEAR_MAX 9999
+
+/* |     Year        |Month   |Day     |Hours   |Minutes |Seconds |Weekday |Fraction|Reason  |
+   |     2 bytes     |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  | = 10 bytes. */
```
```suggestion
/* |     Year        |Month   |Day     |Hours   |Minutes |Seconds |Weekday |Fraction|Reason  |
 * |     2 bytes     |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  | = 10 bytes.
 */
```

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+{
+	int err;
+
+	if (cts_c == NULL || cts_c_init == NULL) {
```
For consistency
```suggestion
	if (!cts_c || !cts_c_init) {
```

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+	int err;
+	struct bt_gatt_read_params *p_read_params;
+
+	err = k_sem_take(&cts_c->read_params_sem, timeout);
```
I wonder if this semaphore is realy needed here. Maybe you could use atomic flag instead it ?
Then you can also remove `bt_gatt_cts_c_on_disconnected`

leewkb4567 replied:
Both semaphore and atomic flag protect the GATT read parameters variable in a similar way. When taking a semaphore, you may specify a timeout which allows waiting a little time. This timing convenience is not provided by atomic flag.

It is found that BLE disconnection does not reset semaphore state. bt_gatt_cts_c_on_disconnected() is written to handle this condition. The use of atomic flag still needs this handling.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+		return -EINVAL;
+	}
+
+	uint32_t index = 0;
```
Please move to the beginning of the function scope or you can simply increment pointer to data `p_data++` instead of creating index variable

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+	}
+
+	/* Month. */
+	if (p_time->exact_time_256.month > 12) {
```
Please define this magic number for this function. Since all this checks return the same error, I think it can be grouped in one if statement

leewkb4567 replied:
Done. Please note the magic number is defined in the start of file.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+ * @retval 0        If the time struct is valid.
+ * @retval -EINVAL  If the time is out of bounds.
+ */
+static int current_time_validate(current_time_char_t *p_time)
```
`p_time` -> `time`

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+ * @retval 0        If the time struct is valid.
+ * @retval -EINVAL 	If the length does not match the expected size of the data.
+ */
+static int current_time_decode(current_time_char_t *p_time,
```
`p_time` -> `time`
`p_data` -> `data`

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2012 - 2020 Nordic Semiconductor ASA
```
I think it could be 2020 only

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+	}
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc,
+					    BT_UUID_CTS_CURRENT_TIME);
+	if (gatt_desc == NULL) {
```
```suggestion
	if (!gatt_desc) {
```

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+	}
+	cts_c->handle_ct = gatt_desc->handle;
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc, BT_UUID_GATT_CCC);
+	if (gatt_desc == NULL) {
```
```suggestion
	if (!gatt_desc) {
```

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+
+	if (!err) {
+		err_code = current_time_decode(&evt.current_time, data, length);
+
```
Unnecessary blank line

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+		} else {
+			/* Verify that the time is valid. */
+			err_code = current_time_validate(&evt.current_time);
+
```
Unnecessary blank line

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_c.c:
```diff
@@ -0,0 +1,251 @@
+{
+	struct bt_gatt_cts_c *cts_c;
+	bt_gatt_cts_c_evt_t evt;
+	int err_code;
```
Can be `err` ?

leewkb4567 replied:
The err variable is already used as ATT error code.

KAGA164 commented on subsys/bluetooth/services/Kconfig:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2019 Nordic Semiconductor ASA
+# Copyright (c) 2019 - 2020 Nordic Semiconductor ASA
```
I think you don't need to modify this year here

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/CMakeLists.txt:
```diff
@@ -1,11 +1,12 @@
 #
-# Copyright (c) 2018 Nordic Semiconductor
+# Copyright (c) 2018 - 2020 Nordic Semiconductor ASA
```
I think you don't need to modify this year here

leewkb4567 replied:
Done.

KAGA164 commented on samples/bluetooth/peripheral_cts_client/src/main.c:
```diff
@@ -0,0 +1,356 @@
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
+};
+
+static char const *day_of_week[] = { "Unknown",	  "Monday",   "Tuesday",
```
`static char const` -> `static const char`

leewkb4567 replied:
Done.

KAGA164 commented on samples/bluetooth/peripheral_cts_client/src/main.c:
```diff
@@ -0,0 +1,356 @@
+	}
+}
+
+static int cts_c_init(void)
```
If you pass event handler as function parameter instead of structure this additional function won't be needed

leewkb4567 replied:
Done.

KAGA164 commented on samples/bluetooth/peripheral_cts_client/src/main.c:
```diff
@@ -0,0 +1,356 @@
+
+	bt_conn_cb_register(&conn_callbacks);
+	if (err) {
+		printk("Failed to register connection callbacks.\n");
```
Remove the dot from log since other `printk` calls in the sample mostly don't print dot

leewkb4567 replied:
Done.

leewkb4567 commented on include/bluetooth/services/cts_c.h:
```diff
@@ -0,0 +1,159 @@
+#endif
+
+/**@brief "Exact Time 256" field of the Current Time characteristic. */
+typedef struct {
```
Done. The PR is rebased to the latest master. The CTS API naming has _gatt removed. Please review.

leewkb4567 commented on include/bluetooth/services/cts_c.h:
```diff
@@ -0,0 +1,159 @@
+};
+
+/**@brief Current Time Service client init structure. This structure contains
+ *        all options and data needed for the initialization of the client. */
```
No longer applicable. This initialization structure is removed.

leewkb4567 commented on include/bluetooth/services/cts_c.h:
```diff
@@ -0,0 +1,159 @@
+ *           Otherwise, a (negative) error code is returned.
+ */
+int bt_gatt_cts_c_init(struct bt_gatt_cts_c *cts_c,
+		       const struct bt_gatt_cts_c_init_param *cts_c_init);
```
Done.

leewkb4567 commented on samples/bluetooth/peripheral_cts_client/src/main.c:
```diff
@@ -0,0 +1,356 @@
+
+	err = bt_gatt_cts_c_handles_assign(dm, &cts_c);
+	if (err) {
+		printk("Could not init CTS client object, error: %d\n", err);
```
The wording is modified to "Could not assign CTS client handles".

leewkb4567 commented on samples/bluetooth/peripheral_cts_client/src/main.c:
```diff
@@ -0,0 +1,356 @@
+	.pairing_failed = pairing_failed
+};
+
+/**@brief Function for handling the Current Time Service errors.
```
When pairing failed, the function calls:

bt_conn_disconnect(conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);

Thus PCA10056 can disconnect from iPhone.

KAGA164 replied:
The function name says that function print current time but doc string says "Function for handling the Current Time Service errors"

leewkb4567 replied:
Done.

KAGA164 commented on subsys/bluetooth/services/cts_client.c:
```diff
@@ -0,0 +1,222 @@
+	time->exact_time_256.fractions256 = *data++;
+
+	/* Reason for updating the time. */
+	time->adjust_reason.manual_time_update = (*data >> 0) & 0x01;
```
Maybe in this line:
`(*data) & BIT(0)`
and next 
`(*data) & BIT(1)`
?

leewkb4567 replied:
Done.
"(*data) & BIT(0)" works.
"(*data) & BIT(1)" does not work as expected. It is modified to "(*data & BIT(1)) ? 1 : 0".

KAGA164 replied:
Sorry about that

KAGA164 commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+
+/**@brief Current Time Service client event handler type. */
+typedef void (*bt_cts_client_evt_handler)(struct bt_cts_client *cts_c,
+					  struct bt_cts_client_evt *evt);
```
`const struct bt_cts_client_evt *evt `?

leewkb4567 replied:
Done.

leewkb4567 commented:
The patch is rebased to v1.4.0-rc2. Checks passed. Please review. Thanks.

joerchan commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+ *
+ * @param[in] cts_c  Current Time Service client structure.
+ */
+void bt_cts_client_on_disconnected(struct bt_cts_client *cts_c);
```
It would be better for the CTS client module to register it's own callback instead of requiring the application to call this function (see BMS)

Possibly this could be removed entirely by removing the need for the semaphore.

leewkb4567 replied:
The bt_conn_cb_register() callbacks have bt_conn structure as parameter. It is not convenient for the CTS client module to get the bt_cts_client structure.

The BAS client tackles this issue by bas_reinit(). It clears the flags on assigning GATT handles. I will follow suit.

joerchan commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+ * @retval 0 If all operations are successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int bt_cts_client_current_time_read(struct bt_cts_client *cts_c,
```
It's preferable to not use the `_client_` infix, and we add it only where we need it (for services), i.e where it would otherwise conflict with a server version of the same function, which is why it is cts_init for server, and cts_client_init for client.

Also read should be place in different order (similar as `bt_bas_read_battery_level`).

So this should be: `bt_cts_read_current_time`

leewkb4567 replied:
Will do.

joerchan commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+ * @retval 0 If the operation is successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int bt_cts_client_handles_assign(struct bt_gatt_dm *dm,
```
`bt_cts_handles_assign`

leewkb4567 replied:
Will do.

joerchan commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+ *          instance of this module.
+ *
+ * @param[in]     dm     Discovery object.
+ * @param[in,out] cts_c  Pointer to the CTS client structure instance for associating the link.
```
` * @param[in] cts_c    Current Time Service client structure.`

leewkb4567 replied:
Will do. Actually it is @param[out] instead.

joerchan commented on subsys/bluetooth/services/cts_client.c:
```diff
@@ -0,0 +1,222 @@
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_cts_client_current_time_read(struct bt_cts_client *cts_c,
+				    k_timeout_t timeout)
```
Did you base this on `bt_hogp_pm_update`?

Usually for these read functions we return -EBUSY instead of queueing using a semaphore.
I would prefer to do that here as well (see bt_hogp_rep_read and bt_bas_read_battery_level for example).

leewkb4567 replied:
Will do to align with other services.

This CTS PR is migrated from nRF5 SDK. I had not given a lot of thought to the design. My initial goal is to make a functional sample.

joerchan commented on subsys/bluetooth/services/cts_client.c:
```diff
@@ -0,0 +1,222 @@
+	const struct bt_gatt_dm_attr *gatt_desc;
+
+	if (bt_uuid_cmp(gatt_service->uuid, BT_UUID_CTS)) {
+		return -ENOTSUP;
```
EINVAL would be more appropriate.

leewkb4567 replied:
The other services return ENOTSUP at this point... You may check bas_client.c, dfu_smp.c, latency_client.c, nus_client.c, throughput.c.

joerchan commented on subsys/bluetooth/services/cts_client.c:
```diff
@@ -0,0 +1,222 @@
+
+	if (bt_uuid_cmp(gatt_service->uuid, BT_UUID_CTS)) {
+		return -ENOTSUP;
+	}
```
Would you mind going through `}` and make sure there is a blank line afterwards, for consistency?

leewkb4567 replied:
Will do.

joerchan commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+
+struct bt_cts_client;
+
+/**@brief Current Time Service client event handler type. */
+typedef void (*bt_cts_client_evt_handler)(struct bt_cts_client *cts_c,
+					  const struct bt_cts_client_evt *evt);
```
I think I'd like to change the event handler to be just a read_cb supplied to the read_current_time function, instead of as an event handler for the module.
Also more often we have an error code in the callback, which could replace the evt_type in this case:
```
int bt_cts_client_init(struct bt_cts_client *cts);

typedef void (*bt_cts_read_cb)(struct bt_cts_client *cts, struct bt_cts_current_time_char *current_time, int err);`

int bt_cts_read_current_time(struct bt_cts_client *cts, bt_cts_read_cb_t func);
```

leewkb4567 replied:
Will do.

joerchan commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+};
+
+/**@brief Data structure for the Current Time characteristic. */
+struct bt_cts_current_time_char {
```
I would drop the `_char` part. `bt_cts_current_time` is good.

leewkb4567 replied:
Will do.

joerchan commented on include/bluetooth/services/cts_client.h:
```diff
@@ -0,0 +1,152 @@
+ */
+int bt_cts_client_handles_assign(struct bt_gatt_dm *dm,
+				 struct bt_cts_client *cts_c);
+
```
Should we have an subscribe function as well? with a notify callback.
Could be used for receiving time adjusts (with the adjust reason set).

leewkb4567 replied:
Will do.

joerchan reviewed (changes requested):
@leewkb4567 Thank you for your contribution. Sorry for the delay in reviewing.

I see you have already updated to align with the recent name update, thank you for that.
Many of the comments I have made is so that we align on how we implement these services.
Using HOGP (previous HIDS_C) as the model is not the best option as this is one of the more complex service/profile.
I would suggest having a look at the battery service client (bas_client.c/h) for comparison.

leewkb4567 commented:
The comments are addressed. Rebased to latest master. Documentation is also updated.
nRF52 boards work fine. Nevertheless, nRF5340 PDK cannot start advertising. A Bluetooth component may be broken...

KAGA164 commented on samples/bluetooth/peripheral_cts_client/src/main.c:
```diff
@@ -0,0 +1,372 @@
+{
+	int err;
+
+	if (has_cts && security_level >= BT_SECURITY_L2) {
```
It seems like you don't need a static variable `security_level`. You could use `bt_conn_get_security` function to check current security level.

leewkb4567 replied:
Done.

leewkb4567 commented on samples/bluetooth/peripheral_cts_client/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+  samples.bluetooth.peripheral_cts_client.build:
+    build_only: true
+    build_on_all: true
+    platform_allow: nrf52dk_nrf52832 nrf52840dk_nrf52840 nrf5340pdk_nrf5340_cpuapp nrf5340pdk_nrf5340_cpuappns
```
Done.

KAGA164 commented on samples/bluetooth/peripheral_cts_client/src/main.c:
```diff
@@ -0,0 +1,368 @@
+{
+	int err;
+
+	if (has_cts && bt_conn_get_security(cts_c.conn) >= BT_SECURITY_L2) {
```
Maybe for readability, please add brackets:` if (has_cts && (bt_conn_get_security(cts_c.conn) >= BT_SECURITY_L2)) `

leewkb4567 replied:
Done.

KAGA164 reviewed (approved):
Looks good. Just one small comment

joerchan commented:
@leewkb4567 This looks good. Can you please rebase onto master and check that it works on nRF5340 PDK, the issue is known and should have been fixed now.

leewkb4567 commented:
Rebased to latest master. nRF5340 PDK works again. Thanks.

joerchan commented:
@leewkb4567 Could you give it one more rebase? Another CI problem happened unfortunately, should be fixed as well.

leewkb4567 commented:
Rebased to latest master. CI passed now.
